src/sql/parts/notebook/models/notebookModel.ts

```diff
@@ -400,25 +412,78 @@ export class NotebookModel extends Disposable implements INotebookModel {
 				this._activeConnection = undefined;
 			}
 
-			clientSession.initialize();
+			await clientSession.initialize();
 			this._sessionLoadFinished = clientSession.ready.then(async () => {
 				if (clientSession.isInErrorState) {
 					this.setErrorState(clientSession.errorMessage);
 				} else {
 					this._onClientSessionReady.fire(clientSession);
 					// Once session is loaded, can use the session manager to retrieve useful info
-					this.loadKernelInfo(clientSession);
+					this.loadKernelInfo(clientSession, this.defaultKernel.display_name);
 				}
 			});
-		});
+		}
+	}
+
+	private updateActiveClientSession(clientSession: IClientSession) {
+		this.clearClientSessionListeners();
+		this._activeClientSession = clientSession;
+		this._clientSessionListeners.push(this._activeClientSession.kernelChanged(e => this._kernelChangedEmitter.fire(e)));
+	}
+
+	private clearClientSessionListeners() {
+		this._clientSessionListeners.forEach(listener => listener.dispose());
+		this._clientSessionListeners = [];
+	}
+
+	public setDefaultKernelAndProviderId() {
+		if (this._savedKernelInfo) {
+			this.sanitizeSavedKernelInfo();
+			let provider = this._kernelDisplayNameToNotebookProviderIds.get(this._savedKernelInfo.display_name);
+			if (provider && provider !== this._providerId) {
+				this._providerId = provider;
+			}
+			this._defaultKernel = this._savedKernelInfo;
+		}
+		else if (this._defaultKernel) {
+			let providerId = this._kernelDisplayNameToNotebookProviderIds.get(this._defaultKernel.display_name);
+			if (providerId) {
+				if (this._providerId !== providerId) {
+					this._providerId = providerId;
+				} else {
+					this._defaultKernel = sqlKernelSpec;
+					this._providerId = SQL_NOTEBOOK_PROVIDER;
+				}
+			}
+		} else if (!this._defaultKernel) {
+			this._defaultKernel = sqlKernelSpec;
+			this._providerId = SQL_NOTEBOOK_PROVIDER;
+		}
+		// update default language
+		this._defaultLanguageInfo = {
```

Comment:
If we're not leveraging version, can we make version optional?

Reply:
@kevcunnane, should we make it optional? I knew Python we supported two versions.